benchmark/benchmark.py

```diff
@@ -0,0 +1,117 @@
+"""Benchmarking utility for PennyLane.
+"""
+# pylint: disable=import-outside-toplevel,invalid-name
+import functools
+import importlib
+import subprocess
+import sys
+
+import numpy as np
+
+import pennylane as qml
+
+
+def timing(func, *, number=10, repeat=5):
+    """Time the given function.
+
+    Args:
+        func (callable[None, Any]): function to time
+        number (int): number of loops per test
+        repeat (int): the number the timing test is run
+    """
+    import timeit
+
+    print('{} loops, {} runs'.format(number, repeat))
+    res = timeit.repeat(func, number=number, repeat=repeat, globals=globals())
+    print("Timing per loop:", np.array(res) / number)
+
+
+def plot(title, kernels, n_range):
+    """Plot timings as a function of the parameter n.
+
+    Args:
+        title (str): plot title
+        kernels (Sequence[callable[Any, Any]]): parametrized benchmarks to time
+        n_range (Sequence[Any]): values the benchmark parameter n takes
+    """
+    import perfplot
+
+    labels = [k.__doc__ for k in kernels]
+    perfplot.show(
+        setup=lambda n: n,
+        kernels=kernels,
+        labels=labels,
+        n_range=n_range,
+        xlabel="n",
+        title=title,
+        # logx="auto",  # set to True or False to force scaling
+        # logy="auto",
+        # equality_check=numpy.allclose,  # set to None to disable "correctness" assertion
+        # automatic_order=True,
+        # colors=None,
+        # target_time_per_measurement=1.0,
+        # time_unit="s",  # set to one of ("auto", "s", "ms", "us", or "ns") to force plot units
+        # relative_to=1,  # plot the timings relative to one of the measurements
+        # flops=lambda n: 3*n,  # FLOPS plots
+    )
+
+
+def profile(func, identifier):
+    """Profile the given function.
+
+    Args:
+        func (callable[None, Any]): function to profile
```

Comment:
So if we want to profile a function that takes arguments, we have to wrap it in another function/lambda? What if we want to profile across different argument values?

Reply:
I see you used `functools.partial` below

Reply:
Profiling across different argument values is possible, but comparing the results is not straightforward. cProfile produces a lot of information, which we here save in a file. The profile stats are also printed, but the idea is for the user to take a better look at the stats using some sort of viewing tool. I added the script `view_pstats.sh`for viewing the results in the form of a call graph, which is more informative.